scripts/release.py

```diff
@@ -0,0 +1,244 @@
+"""
+Script for automating changes necessary for `ruff-vscode` releases.
+
+This script does the following things:
+- Bumps the version of this project in `pyproject.toml` and `package.json`
+- Bumps the `ruff` and `ruff-lsp` dependency pins in `pyproject.toml`
+- Updates the changelog and README
+- Updates the package's lockfiles
+"""
+
+# /// script
+# requires-python = "==3.7.*"
+# dependencies = ["packaging", "requests", "rich-argparse", "tomli", "tomlkit"]
+#
+# [tool.uv]
+# exclude-newer = "2024-11-27T00:00:00Z"
+# ///
+from __future__ import annotations
+
+import argparse
+import json
+import re
+import subprocess
+import textwrap
+from dataclasses import dataclass
+from pathlib import Path
+
+import requests
+import tomli
+import tomlkit
+import tomlkit.items
+from packaging.requirements import Requirement
+from packaging.specifiers import SpecifierSet
+from packaging.version import Version
+from rich_argparse import RawDescriptionRichHelpFormatter
+
+PYPROJECT_TOML_PATH = Path("pyproject.toml")
+PACKAGE_JSON_PATH = Path("package.json")
+README_PATH = Path("README.md")
+CHANGELOG_PATH = Path("CHANGELOG.md")
+
+
+@dataclass(frozen=True)
+class RuffVersions:
+    existing_vscode_version: Version
+    new_vscode_version: Version
+    existing_ruff_pin: Version
+    latest_ruff: Version
+    existing_ruff_lsp_pin: Version
+    latest_ruff_lsp: Version
+
+
+def existing_dependency_pin(
+    dependencies: dict[str, SpecifierSet], dependency: str
+) -> Version:
+    """Return the version that `dependency` is currently pinned to in pyproject.toml."""
+    specifiers = dependencies[dependency]
+    assert len(specifiers) == 1
+    single_specifier = next(iter(specifiers))
+    assert single_specifier.operator == "=="
+    return Version(single_specifier.version)
+
+
+def latest_pypi_version(project_name: str) -> Version:
+    """Determine the latest version of `project_name` that has been uploaded to PyPI."""
+    pypi_json = requests.get(f"https://pypi.org/pypi/{project_name}/json")
+    pypi_json.raise_for_status()
+    return Version(pypi_json.json()["info"]["version"])
+
+
+def get_ruff_versions() -> RuffVersions:
+    """
+    Obtain metadata about the project; figure out what the new metadata should be.
+    """
+    with PYPROJECT_TOML_PATH.open("rb") as pyproject_file:
+        pyproject_toml = tomli.load(pyproject_file)
+
+    existing_vscode_version = pyproject_toml["project"]["version"]
+
+    major, minor, micro = Version(existing_vscode_version).release
+    new_vscode_version = Version(f"{major}.{minor + 2}.{micro}")
```

Comment:
Hmm, how would we do a patch release or release a preview version (which we've done a few times in the past)?

Reply:
My idea of this script was to cover the _common_ workflow rather than _all possible_ workflows. I can add a CLI flag for this if you want, but I worry that the complexity it will add to the script won't be worth the time it will save us

Reply:
The problem I see is that this PR removes the documentation on how to do a release manually when not using the script. To me, that means it's now the only way of doing a release. 

That's why I think we should either revert the documentation changes or extend the script. I'm worried about reverting the documentation because it probably will get outdated with future changes to this script which then is even worse: You end up with incomplete/incorrect instructions

Reply:
I see. Okay, I'll add the CLI flags.

Reply:
I added more flags in https://github.com/astral-sh/ruff-vscode/pull/647/commits/37b29352724609e8f8fa1926e4043905fea2376d, and tested them using `uv run --python=3.7 scripts/release.py --new-version=2024.600.0 --new-ruff=0.6.0`. It did everything I expected it to.

Also, it didn't end up being much more code at all, so I think it _was_ worth it -- thanks for persuading me!

Reply:
Thanks for adding it and thanks for automating the steps!